### PR TITLE
Admin: Include a "None" option for calculator type user in shipping method and payment method forms

### LINKED ISSUE #9424: Calculator: Provide a calculator 'none'
## Description
**- As a:** enterprise user or super admin
**- On page:** `/admin/shipping_methods/new` `/admin/shipping_methods/XXX/edit` and  `/admin/payment/new` `/admin/payment/XXX/edit` 
**- I want to be able to do:** set the calculator to "none" or "empty" by default.

## Acceptance Criteria & Tests

1. Log in as enterprise user
2.  Go to `/admin/shipping_methods/new` `/admin/shipping_methods/XXX/edit` and  `/admin/payment/new` `/admin/payment/XXX/edit` 
3. See that when creating the dropdown of the calculator is set to "none" by default, and that you can change it to "none" when editing.
4. When using "none" the calculator has no value

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

This PR creates a `None` calculator for both shipping method and payment method.

Capture d’écran 2022-12-27 à 15 26 14

Closes #9424

#### What should we test?

- As an admin, check that payment methods and shipping methods doesn't need to set a calculator and can be set to none. 
- check that this None calculator is well handled in each part of the app where they should be. 
- by default, when creating a new shipping method or payment method, calculator should be set to `None`

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Excellent.

**Comment by drummer83** [comment]
Hi @jibees,
Thanks for this one!

I created the following test cases:

| Page | (Default) value **before** PR | (Default) value **after** PR | Properly working on checkout | Properly working in reports |
| - | - | - | - | - | 
| /admin/shipping_methods/new | flat percent | flat percent :x: | ? | ? |
| /admin/shipping_methods/XXX/edit | flat percent 0.0 | flat percent 0.0 | ? | ? |
| /admin/payment/new | flat rate (per order) | flat rate (per order) :x:  | ? | ? |
| /admin/payment/XXX/edit | flat rate (per order) 10.0 | flat rate (per order) 10.0  | ? | ? |

According to the description and acceptance criteria in the original issue, the calculator 'none' should be the default when creating new payment/shipping methods. This is currently not the case.

Also I think that 'None' should always be at the top of the list in the drop down.

image

I did not continue testing at this point.
Could you fix these before testing again and merging? 

Thanks! I will try to move to In Dev - if that's possible at the moment...

**Comment by jibees** [comment]
> the calculator 'none' should be the default when creating new payment/shipping methods

I thought I'd a good reason to not do this, but I can't remember. So will implement it.


> Also I think that 'None' should always be at the top of the list in the drop down.

It's sorted alphabetically. Not sure 'None' should be on top then. But when not!

**Comment by jibees** [comment]
Many errors: 

```
 PG::NotNullViolation: ERROR:  null value in column "calculable_id" violates not-null constraint
            DETAIL:  Failing row contains (6, Calculator::None, null, Spree::ShippingMethod, 2023-01-10 10:57:24.567727, 2023-01-10 10:57:24.567727).

```

**Comment by jibees** [comment]
This has changed, so request a code review from @mkllnk again.

Thanks! 🙏

**Comment by abdellani** [comment]
As we discussed before, the only question that I have is: why do we need to check `new_record?` on the `ShippingMethod#init` while this is not needed on  `PaymentMethod#init`?
```ruby
    def init
      self.calculator ||= ::Calculator::None.new if new_record?
    end
```
Otherwise, it looks good to me.

Thannk you @jibees  :+1:

**Review by abdellani — APPROVED** [review]

File: app/models/spree/shipping_method.rb
Comment (mkllnk) [inline app/models/spree/shipping_method.rb:38]: I would have put this hook above `after_save` to have a chronological order.

Comment (mkllnk) [reply app/models/spree/shipping_method.rb:38]: I never used this callback before. I hope it works as expected. Calling this in the controller would not have been enough? The fewer callbacks the better.

**Review by mkllnk — APPROVED** [review]

File: app/models/spree/shipping_method.rb
Comment (mkllnk) [reply app/models/spree/shipping_method.rb]: Okay, good for consistency.

**Review by mkllnk — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @jibees ,

After this PR,

- the default calculator when creating a new shipping or payment method is the new `None` calculator  :green_circle: 

I've checked out with these methods, and spotted no regressions.

The app handles the case in which a shipping tax category is associated to a None calculator correctly  :green_circle: 

image

The adjustments table shows the tax correctly  :green_circle: 

image

Changing shipment and payment works well - while editing an order :green_circle: 

I've checked several reports, namely
- Delivery Report
- Orders and Distributors
- Payments reports

The methods with None calculator are displayed correctly.

Looks good to go! Merging.

**Comment by filipefurtad0** [comment]
Hey @jibees ,

I'm testing something else - long after merging this PR - in staging-AU, but somehow, I can't observe this option:

image

This is really strange, as I we're sure that staging a PR **always deploys a version of that PR with current master**.  I don't know why, but this did not happen here.

A few deploys later, and I can reproduce the real concerning bit:

The Stripe bump seems to throw this error:
```

Unhandled error
ActionView::Template::Error: The single-table inheritance mechanism failed to locate the subclass: 'Calculator::None'. This error is raised because the column 'type' is reserved for storing the class in case of inheritance. Please rename this column if you didn't intend it to be used for storing the inheritance class or overwrite Spree::Calculator.inheritance_column to use another column for that information.
```
taken from: https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/63dab1ca207da80008822f7c?event_id=63dab1ca00ad99ac8b960000&i=sk&m=nw

I've rebased the PR from the [Stripe bump](https://github.com/openfoodfoundation/openfoodnetwork/pull/10269) to check whether our specs actually catch this.

I'm posting here as I'm assuming there is some relation to this PR (`Calculator::None` is mentioned).

**Comment by jibees** [comment]
>  PR always deploys a version of that PR with current master

This is something I'm also convinced but actually I don't have any argument, and maybe this should be investigate to be sure about that assertion.

**Pull request merged**
